zebra/zserv.c

```diff
@@ -441,7 +442,12 @@ static int zserv_read(struct thread *thread)
 			   zebra_route_string(client->proto));
 
 	/* Reschedule ourselves */
-	zserv_client_event(client, ZSERV_CLIENT_READ);
```

Comment:
How about checking the socket pending bytes instead of blindly asking for reads?

(please read comment at the bottom of this comment)

Example (man 7 tcp):

```
   Ioctls
       The following ioctl(2) calls return information in value.  The
       correct syntax is:

              int value;
              error = ioctl(tcp_socket, ioctl_type, &value);

       ioctl_type is one of the following:

       SIOCINQ
              Returns the amount of queued unread data in the receive
              buffer.  The socket must not be in LISTEN state, otherwise an
              error (EINVAL) is returned.  SIOCINQ is defined in
              <linux/sockios.h>.  Alternatively, you can use the synonymous
              FIONREAD, defined in <sys/ioctl.h>.
```

I have the same problem with BFD daemon: if you have a peer with high volume of control packets we end up spending too much time on `poll()` instead of actual `read()` work. I think that each case might have its own sweet spot, but here are some things to consider:

1. First of all: always measure.
2. Your case might be able to benefit from `SIOCINQ`, however BFD's case not. It might also end up in the same spot because of syscalls, maybe a simple buffer size bump will help (see item 4).
3. BFD's case might get better if we use `recvmmsg` (receive multiple packets with a single `read()`), however it might benefit even more if we could get UDP packets queue (`SIOCINQ` only returns the number of next packet length, but we could find a way to extract this information from `/proc/net/udp`). -- Considering your solution would also apply to BFD.
4. It is also possible that by doing more `read()` work instead of distributing with other fds from `poll()` we might starve other "threads"/"functions", but then we come back at the problem of "wasting" time in `poll()`.

Reply:
Thanks for the thoughts: yes, this is not a zebra-specific problem / pattern! A couple of things:

1, 2: yes, but my concern as you say is that we'd still be doing system calls when we have a very reasonable hint that there was more data that we could read - because we had exhausted our specific limit on number of packets per read iteration.

4: no, in this specific case this pthread _only_ reads and writes packets; this is not the zebra main pthread that's doing lots of other kinds of work. there isn't really anything else that can be starved (imo).

I was trying to simplify the case where we're receiving a burst of messages from a daemon, and the pthread is spending time diving into and out of poll() when it has a very strong hint that it has data available to read on its socket.